src/codegen/compile_classical.ml

```diff
@@ -10928,18 +10967,23 @@ and compile_prim_invocation (env : E.t) ae p es at =
          StackRep.of_arity return_arity,
 
          code1 ^^
-         compile_unboxed_zero ^^ (* A dummy closure *)
+         Type.(match as_obj par.note.Note.typ with
+               | Object, [] -> compile_unboxed_zero (* a dummy closure *)
+               | _ -> compile_exp_vanilla env ae par) ^^ (* parenthetical *)
          compile_exp_as env ae (StackRep.of_arity n_args) e2 ^^ (* the args *)
          G.i (Call (nr (mk_fi ()))) ^^
          FakeMultiVal.load env (Lib.List.make return_arity I32Type)
       | _, Type.Local ->
-         let (set_clos, get_clos) = new_local env "clos" in
+         let set_clos, get_clos = new_local env "clos" in
 
          StackRep.of_arity return_arity,
          code1 ^^ StackRep.adjust env fun_sr SR.Vanilla ^^
+         Closure.prepare_closure_call env ^^ (* FIXME: move to front elsewhere too *)
          set_clos ^^
-         get_clos ^^
-         Closure.prepare_closure_call env ^^
+         Type.(match as_obj par.note.Note.typ, ret_tys with
+               | (Object, []), _ -> get_clos (* just the closure *)
+               | _, [ret] when is_async_fut ret -> Arr.lit env Tagged.T [compile_exp_vanilla env ae par; get_clos] (* parenthetical: pass a pair *)
```

Comment:
`is_async_fut` is the wrong predicate! When `T -> async U` `Local` types get lowered, we end up with `(,,) -> { #schedule; #suspend }`